pkg/controller/replicaset/replica_set.go

```diff
@@ -650,3 +629,74 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
 	}
 	return manageReplicasErr
 }
+
+// slowStartBatch obtains total batch size to be processed and a function to execute during each batch.
+// It returns successfully processed batch size and an error if found.
+func slowStartBatch(totalBatchSize int, fn func() error) (int, error) {
+	var wg sync.WaitGroup
+	remainedBatchSize := totalBatchSize
+	totalProcessedBatchSize := 0
+	errCh := make(chan error, totalBatchSize)
+	var lock sync.Mutex
+	for batchSize := integer.IntMin(remainedBatchSize, controller.SlowStartInitialBatchSize); remainedBatchSize > 0; batchSize = integer.IntMin(2*batchSize, remainedBatchSize) {
+		wg.Add(batchSize)
+		oldTotalProcessedBatchSize := totalProcessedBatchSize
+		for i := 0; i < batchSize; i++ {
+			go func() {
+				defer wg.Done()
+				err := fn()
+				if err != nil && !errors.IsTimeout(err) {
+					errCh <- err
+				} else if err == nil {
+					lock.Lock()
+					totalProcessedBatchSize++
+					lock.Unlock()
+				}
+
+			}()
+		}
+		wg.Wait()
+		// The skipped pods will be retried later. The next controller resync will
+		// retry the slow start process.
+		curProcessedBatchSize := totalProcessedBatchSize - oldTotalProcessedBatchSize
+		if curProcessedBatchSize < batchSize {
+			return curProcessedBatchSize, <-errCh
+		}
+		remainedBatchSize -= batchSize
+	}
+	return totalBatchSize, nil
```

Comment:
Yes, with the suggested new changes, this may be different from `count` for corner cases (e.g., initialBatchSize<=0), so we should return `successes`.